bootstrap/pkg/kfapp/coordinator/coordinator.go

```diff
@@ -767,6 +769,13 @@ func LoadKfAppCfgFile(cfgfile string) (kftypesv3.KfApp, error) {
 		}
 	}
 
+	if err := kfdef.SyncCache(); err != nil {
```

Comment:
I think this is the common entrypoint function that everyone uses to load a KfApp.
Will look into it again.

Reply:
That might be the case but I think the better pattern would be to call SyncCache near the code that depends on it rather than relying on it already being invoked.

e.g. in kustomize.go generate we call SyncCache
https://github.com/kubeflow/kubeflow/blob/2395c5fc0d384cf6dd4d7167bbe9010cd1315b54/bootstrap/pkg/kfapp/kustomize/kustomize.go#L420

So arguably it should be the Arrikto plugin that invokes SyncCache before accessing the cache.
Calling SyncCache should be a null op if the cache is up to date.

Reply:
+1 `SyncCahe()` should be called in the respective plugins

Reply:
> Calling SyncCache should be a null op if the cache is up to date.

I see that calling SyncCache right now is indeed cheap, however that might change in the future when we find a better way to check if the cache is up-to-date.
The current way just checks for the directories.

Personally, I don't believe it's unreasonable for Plugins to expect that caches have been synced before their functions are called (`GetPlatform`, `Init`, `Generate`, `Apply`).
Otherwise, every Plugin needs to make sure to call SyncCache() before running each of these functions, leading to a lot of duplication. In addition, if the SyncCache() operation becomes more expensive in the future, it will be a great slowdown.

I propose calling `SyncCache()` in coordinator's `GetPlatform`, `Init`, `Generate` and `Apply`, where we know that Plugin functions will be called next.
WDYT?

Reply:
Calling it in Generate and Apply seems reasonable.

Reply:
Delete also needs it, or will need it in the future I believe.
GetPlatform also depends on it for existing_arrikto, in order to get the cache's LocalPath.
Are you ok with making a contract of:

Whenever a Plugin exposed function (GetPlatform, Init, Generate, Apply, Delete) is performed, the cache is synced beforehand.
This is what I have implemented now.

This is also blocking the existing_arrikto e2e test (https://github.com/kubeflow/kubeflow/pull/4154).

Reply:
Why do we need to create new contract right now? The code will continue to be refactored (e.g. GetPlatform should probably go away given we no longer have platforms): how would we enforce this contract?

SyncCache is cheap right now and the simple solution for ensuring SyncCache is run before code that depends on it to invoke SyncCache.

Trying to minimize calls to SyncCache feels like premature optimization at this point.

My suggestion is to just call SyncCache in any code inside your plugin that depends on it being uptodate.

If you are worried about the efficiency of SyncCache file a P2 bug to improve it later.

Reply:
@jlewi fair enough.
I left SyncCache only in Generate, Apply and Delete.
Also inside existing_arrikto where it is needed.